### PR TITLE
[Impeller Scene] Add DisplayList OP and Dart bindings

### PR DESCRIPTION
**Pull request by bdero**
This lays the ground work for building experimental canvas-backed Scene widgets alongside the framework.

The intention is to enable the following:
1. Creation of nodes from ipscene assets.
2. Creation of empty nodes with specific transforms.
3. Parenting nodes to each other (and allow nodes to have multiple parents).
4. Drawing nodes via a color source.
5. Safely overwriting transforms for existing nodes.
6. Basic control of animation clips (play/pause, seek, weights/blending).

1-4 are already covered (although actual ipscene assets aren't bundled yet, of course). For 5 and 6, a simple mutation log will be pushed to the `DlColorSource` and applied on the raster thread prior to rendering (good suggested design by Chinmay).

Every UI `SceneNode` is backed by an Impeller Scene `Node`, but not every Impeller Scene `Node` belongs to a UI `SceneNode`; Imported ipscenes are arbitrarily complex scene descriptions, but only the imported scene's root gets referenced by a `SceneNode`.

Screenshot of the DL playground:
image

### CODE REVIEW

**Review by jonahwilliams — COMMENTED** [review]
If the scene is incoporated into the ... other scene via a color source, I'm almost certain you'd have to drive all animations via the UI thread instead of the raster thread. 

As long as we don't raster cache ever on impeller, it should still work, but the separate layer was needed for textures on the Skia backend so we could more easily disable raster caching of parent layers.

**Comment by bdero** [comment]
> If the scene is incoporated into the ... other scene via a color source, I'm almost certain you'd have to drive all animations via the UI thread instead of the raster thread.

For animation behavior that the user is defining inline in the framework, yes. This is a bit of a learning exercise for me, so if there's animation driven by the raster thread in Flutter today, I'd like to take a look -- especially if it doesn't require repaints. :)

For externally authored animations embedded in the asset (which will be incredibly intricate), high level controls are triggered on the UI thread and the Node's animation player drives it on the raster thread just before rendering.

It seemed like this would be simple and fast enough for now.

> As long as we don't raster cache ever on impeller, it should still work, but the separate layer was needed for textures on the Skia backend so we could more easily disable raster caching of parent layers.

Hmm, do CustomPainters get raster cached without additional intervention?

**Comment by jonahwilliams** [comment]
> For animation behavior that the user is defining inline in the framework, yes. This is a bit of a learning exercise for me, so if there's animation driven by the raster thread in Flutter today, I'd like to take a look -- especially if it doesn't require repaints. :)

I'm not sure what we do for external textures, I would start there. Something like video player? I hope we're not driving every frame of a video through the UI thread...

> Hmm, do CustomPainters get raster cached without additional intervention?

Pretty much anything can get raster cached based on complexity scoring, this caching is not very content aware and can lead to all sorts of fun bugs, like raster caching a picture with a blend mode that depends on the dest - causing the apparent color to change based on whether it is draw in the raster cache picture or normally.

I don't think we should worry about this case because we don't have raster caching.

**Comment by bdero** [comment]
Another appropriate analogy is 2D games today. If someone makes a realtime 2D platformer game powered by Flutter, on which thread will the user compute the character's position for display on the next frame? My understanding is that this will always be done on the UI thread today, as the UI thread stages render instructions for the next frame.

**Comment by jonahwilliams** [comment]
Discussed offfline, but driving via the UI thread is fine for now

File: impeller/display_list/BUILD.gn
Comment (chinmaygarde) [inline impeller/display_list/BUILD.gn:30]: Why behind a flag?

Comment (jonahwilliams) [reply impeller/display_list/BUILD.gn:30]: When I added the stubs originally, I believe there were some lint complaints about unused code.

Comment (bdero) [reply impeller/display_list/BUILD.gn:30]: Fuchsia doesn't build the shaders right now, so I ended up guarding everything to avoid the Fuchsia build depending on Impeller shader headers (Nodes depend on Materials, Materials depend on the shader headers).

Comment (bdero) [reply impeller/display_list/BUILD.gn:30]: A cleaner solution would probably be to have the DL op and friends use agnostic type (e.g. DlSceneNode) and compile an empty version that errors out on use if 3D support is off.

**Comment by bdero** [comment]
I'm gonna add a DL test before landing, but this should be good to go after all of yesterday's fixes. I spent too long trying to add an observatory test that runs when the 3d flag is enabled (see the reverts), but it was just way too much hackery that needs to be undone later.

I'm working on the mutator log, but would like to add it in a separate patch.

File: lib/ui/natives.dart
Comment (jonahwilliams) [inline lib/ui/natives.dart:86]: You can't remove these from the regular UI build.

Comment (jonahwilliams) [reply lib/ui/natives.dart:86]: maybe add a separate setup method for the 3d scene and stub it in the regular version?

Comment (bdero) [reply lib/ui/natives.dart:86]: It's still in the regular UI build via https://github.com/flutter/engine/pull/38676/files#diff-59c10ede29c01c127b31d48f6ab7ffb2633dbe85b0831cab377cafc9e389156bR41

Comment (jonahwilliams) [reply lib/ui/natives.dart:86]: ahh got it.

File: lib/ui/painting/scene/scene_node.cc
Comment (bdero) [inline lib/ui/painting/scene/scene_node.cc:96]: @chinmaygarde Could you spot check the threading code above? 

I pulled the dart state persistence stuff from the image decoding routine when I ran into trouble with it. I think I know how it works now, but I'm not super experienced with the Dart runtime's behavior.

Also, I'm fetching the Impeller context from an IO ref because that's all I happen to have found. Is there an easily accessible ref I can grab on the raster thread instead?

**Comment by bdero** [comment]
Added a DL playground.

File: lib/ui/setup_hooks.dart
Comment (chinmaygarde) [inline lib/ui/setup_hooks.dart:15]: We should document these service protocol extensions in https://github.com/flutter/flutter/wiki/Engine-specific-Service-Protocol-extensions.

File: lib/ui/painting/scene/scene_shader.cc
Comment (chinmaygarde) [inline lib/ui/painting/scene/scene_shader.cc:51]: "There's no way". Here and elsewhere, perhaps file bugs. But if this is for a demo, we can do a pass later to rethink this.

**Review by chinmaygarde — APPROVED** [review]

File: lib/ui/painting/scene/scene_node.cc
Comment (chinmaygarde) [reply lib/ui/painting/scene/scene_node.cc:96]: So, I am not sure why the scene needs to be created on the raster task runner. That is just going to cause jank. We should decompress it on the concurrent worker pool like the texture decompression. But this is fine for now and safe.

Comment (bdero) [reply lib/ui/painting/scene/scene_node.cc:96]: Huh, I guess I haven't verified the safety of doing it off the raster thread. In the moment I thought the geometry + texture buffer uploads would fall over for GLES because we don't manage multiple contexts... but actually I'm pretty sure that's not an issue because IIRC we enqueue everything in the reactor and then execute it on the raster thread later anyways.

File: display_list/display_list.h
Comment (flar) [inline display_list/display_list.h:166]: It seems like a lot of these ifdef blocks could be eliminated in favor of a single ifdef that controls whether or not the FOR_EACH macro contains `V(SetSceneColorSource)` If that `V()` entry doesn't exist then the code isn't generated for that op (and vice versa).

File: lib/ui/painting/scene/scene_node.cc
Comment (chinmaygarde) [reply lib/ui/painting/scene/scene_node.cc:96]: Yeah. I want to move away from the model of certain objects being only safely accessible from specific threads. If there is missing synchronization, that is a definitely a construction issue and not one that should restrict the use of the object from any thread.

You're right about the OpenGL ES stuff using the reactor.

**Pull request merged**
